In [1]:
# ! pip install pycaret

# Credit Card Fraud Detection with PyCaret

In this notebook, we will test out the autoML [PyCaret](https://pycaret.org/) with reference to its [Binary Classification tutorial](https://github.com/pycaret/pycaret/blob/master/tutorials/Binary%20Classification%20Tutorial%20Level%20Beginner%20-%20%20CLF101.ipynb). Using PyCaret, we will select and train a few models to train the data generated by the [synthetic credit card transactions generator](./01-Synthetic-data-generation.ipynb).  

Our ultimate goal is to select a couple of models and use the trained models to perform real-time credit card fraud detection.  
Combined with [atoti](https://atoti.io/), we will compute the impact on key business metrics and also compare the performance of the models.

In [2]:
import glob
from datetime import date
from zipfile import ZipFile

import numpy as np
import pandas as pd
import wget
from haversine import Unit, haversine
from IPython.display import clear_output, display
from utils import data_prep

## 1 Data preparation

### 1.1 Gather customer and credit card transactions data  

You may use the [01-Synthetic-data-generation](./01-Synthetic-data-generation.ipynb) notebook to generate your own data. Alternatively, you can download a saved version of customers and transactions from Jun - Oct 2021.

In [3]:
data_path = "./output"

Skip the below step if data has already been downloaded.

In [4]:
# def bar_custom(current, total, width=80):
#     clear_output(wait=True)
#     print("Downloading: %d%% [%d / %d] bytes" % (current / total * 100, current, total))


# url = "https://data.atoti.io/notebooks/credit-card-fraud/output.zip"
# filename = wget.download(url, bar=bar_custom)

# # unzipping the file
# with ZipFile("output.zip", "r") as zipObj:
#     # Extract all the contents of zip file in current directory
#     zipObj.extractall(data_path)

#### 1.1.1 Process transactions files

We concatenate the transactions for all the profiles into a single DataFrame.

In [5]:
all_files = glob.glob(data_path + "/*adults*.csv")

In [6]:
files = []

for filename in all_files:
    df = pd.read_csv(filename, index_col=None, header=0)
    files.append(df)

txns_df = pd.concat(files, axis=0, ignore_index=True)
print(
    "Total number of transactions: ",
    len(txns_df),
    " - number of columns: ",
    len(txns_df.columns),
)

txns_df.head()

Total number of transactions:  1276543  - number of columns:  14


ssn            cc_num     acct_num                        profile  \
0  028-33-9309  4391470058239421  76479233103  adults_2550_female_rural.json   
1  028-33-9309  4391470058239421  76479233103  adults_2550_female_rural.json   
2  028-33-9309  4391470058239421  76479233103  adults_2550_female_rural.json   
3  028-33-9309  4391470058239421  76479233103  adults_2550_female_rural.json   
4  028-33-9309  4391470058239421  76479233103  adults_2550_female_rural.json   

                          trans_num  trans_date trans_time   unix_time  \
0  a32d15100140b93f67430a0d51f7f470  2021-08-24   03:09:54  1629774594   
1  4302c8e427aebde7200b8cb312b092e6  2021-08-23   02:35:13  1629686113   
2  3e2edc0fa18cb831ab946b3636b2beb4  2021-08-24   23:16:43  1629847003   
3  a33a6497109e098aa322c904f105d81b  2021-08-24   22:30:43  1629844243   
4  c902dfaa64a6756f7d0f6e5fbea02284  2021-08-23   01:03:58  1629680638   

        category      amt  is_fraud                          merchant  \
0  gas_transport     8.78         1             fraud_Prohaska-Murray   
1    grocery_pos   368.16         1           fraud_Cartwright-Harris   
2           home   276.14         1              fraud_Beier and Sons   
3    food_dining   106.20         1  fraud_Kutch, Steuber and Gerhold   
4   shopping_net  1094.50         1                    fraud_Jast Ltd   

   merch_lat  merch_long  
0  43.469516  -85.860994  
1  44.040299  -84.026526  
2  43.735269  -85.122360  
3  43.432890  -85.804933  
4  43.245551  -85.220724

#### 1.1.2 Reducing number of merchants 

In the synthetic data generator, merchants' location are computed based on distance from customer. Hence, there may be many merchants of the same name close by. We are going to clean up by combining the same merchants that are close by.

In [7]:
txns_df = data_prep.merchant_cleanup(txns_df)

In [8]:
txns_df.sort_values(by=["trans_date", "trans_time"], inplace=True)
txns_df.head()

ssn               cc_num      acct_num  \
350538   651-85-6277     6011254895433784  650277300926   
156578   829-65-8471     3561104445052090  359048776675   
1030583  765-51-3253       30177657960332  457496356031   
473195   843-90-9424  4893518176925451756  261496133542   
337806   201-15-2262     3510018921883087  840477491610   

                               profile                         trans_num  \
350538     adults_2550_male_rural.json  78a301f6c1d629851f2134f3d428bd64   
156578   adults_2550_female_urban.json  b1f41c9941a8fef0c58b7da9c53b87d1   
1030583    adults_50up_male_urban.json  4c277b5b47525ffd52919af6c80f292a   
473195     adults_2550_male_urban.json  a24a9cfd2eb04f1e477c8b2b28f57bd7   
337806     adults_2550_male_rural.json  2ecedf1776c3de3f8e87cd4842f855f4   

         trans_date trans_time   unix_time     category    amt  is_fraud  \
350538   2021-06-01   00:03:16  1622505796  grocery_pos  94.07         0   
156578   2021-06-01   00:05:54  1622505954     misc_net   2.52         0   
1030583  2021-06-01   00:06:50  1622506010  food_dining  40.44         0   
473195   2021-06-01   00:09:07  1622506147  grocery_pos  82.00         0   
337806   2021-06-01   00:10:19  1622506219  grocery_pos  82.95         0   

                     merchant  merchant_id  merch_long  merch_lat  
350538       fraud_Barton Inc       191095  -80.583026  34.505328  
156578    fraud_Ruecker Group       156108  -74.303208  40.728483  
1030583     fraud_Haag-Blanda       289878  -87.214427  42.418391  
473195     fraud_Bauch-Raynor       418020  -97.492732  36.430343  
337806   fraud_Kiehn-Emmerich       337806  -76.540490  43.925550

In [9]:
txns_df.shape

(1276543, 15)

In [10]:
print(
    f"Non-fraudulent: {len(txns_df[txns_df['is_fraud']==0])}, Fraudulent: {len(txns_df[txns_df['is_fraud']==1])}"
)

Non-fraudulent: 1185276, Fraudulent: 91267


#### 1.1.3 Load customer file  

In [11]:
cust_df = pd.read_csv(f"{data_path}/customers_list.csv")
cust_df.head()

ssn            cc_num     first        last gender  \
0  535-89-6775   346545733014127     Carla    Sheppard      F   
1  354-81-2556  3518152160915706     Jaime       Boyle      F   
2  422-27-9393  5429874770003176      Seth    Mitchell      M   
3  028-33-9309  4391470058239421  Danielle   Rodriguez      F   
4  246-23-6185  5395351866032237    Nicole  Villarreal      F   

                            street        city state    zip      lat  \
0  013 Daniel Trafficway Suite 429  Pittsburgh    PA  15213  40.4440   
1    88361 Schaefer Cove Suite 452      Lester    AL  35647  34.9596   
2                 79740 Maria View      Santee    CA  92071  32.8486   
3     8929 Kenneth Courts Apt. 591        Muir    MI  48860  43.0439   
4             2655 Rebecca Parkway     Welches    OR  97067  45.3399   

       long  city_pop                         job         dob      acct_num  \
0  -79.9552    687276  Psychologist, occupational  1938-02-10  438882350784   
1  -87.1006      1030          Theme park manager  1944-10-08  347776380945   
2 -116.9862     53422      Engineer, agricultural  1966-07-13   97717754818   
3  -84.9391      1334   Publishing rights manager  1992-04-12   76479233103   
4 -121.9598      2138   Education officer, museum  1999-08-22  389654813660   

                          profile  
0   adults_50up_female_urban.json  
1   adults_50up_female_rural.json  
2     adults_50up_male_urban.json  
3   adults_2550_female_rural.json  
4  young_adults_female_rural.json

In [12]:
len(cust_df)

9353

#### 1.1.4 Compute additional features

The `compute_features` function merge the customers and transactions to compute additional features:
- customer age
- distance between customer and merchant
- time of transaction
    - Weekday or weekend
    - Day or night
- cumulative number of transactions for customer in the past 1, 7 and 30 days
- cumulative average transaction amount for customer in the past 1, 7 and 30 days

In [13]:
cc_txn_df = data_prep.compute_features(cust_df, txns_df)

In [14]:
len(cc_txn_df)

1276543

In [15]:
cc_txn_df.tail()

ssn            cc_num      acct_num  \
1276538  402-78-5680    30306133731849  683914675882   
1276539  590-91-7122  3525276120447046  301548051587   
1276540  403-93-5583      639082412268  847752784845   
1276541  329-13-2201    30179130072305  205446777871   
1276542  790-64-5257  3552783966051994  573043127801   

                               profile                         trans_num  \
1276538    adults_50up_male_urban.json  e0570d36e082a5913159e90459f95763   
1276539   young_adults_male_urban.json  5087d30049ca44c3a8b06f5f16d4400a   
1276540  adults_2550_female_rural.json  a6f360bf3f1eee039d16a3dffc7caab0   
1276541  adults_50up_female_urban.json  e2597d6acf0b135cd9e261c257051bb1   
1276542  adults_50up_female_rural.json  91ae1f2cd55e7d21d4059c2d3bc56bd1   

        trans_date trans_time   unix_time       category     amt  ...  age  \
1276538 2021-10-31   23:58:08  1635724688       misc_net  780.88  ...   74   
1276539 2021-10-31   23:58:16  1635724696   shopping_net  900.05  ...   24   
1276540 2021-10-31   23:58:41  1635724721         travel   10.29  ...   45   
1276541 2021-10-31   23:59:55  1635724795       misc_net  891.91  ...   75   
1276542 2021-10-31   23:59:58  1635724798  entertainment  610.12  ...   81   

        distance_from_cust  txn_during_night  trans_weekend  nb_txns_1_days  \
1276538          87.147193                 1              1             7.0   
1276539          69.105396                 1              1             8.0   
1276540         126.952035                 1              1             6.0   
1276541          98.274275                 1              1             7.0   
1276542         234.908522                 1              1             4.0   

        avg_txns_amt_1_days nb_txns_7_days avg_txns_amt_7_days  \
1276538          562.718571           12.0          553.205833   
1276539          466.542500           17.0          546.149412   
1276540          412.798333           15.0          312.095333   
1276541          400.428571           12.0          446.110000   
1276542          510.600000            7.0          297.687143   

        nb_txns_30_days avg_txns_amt_30_days  
1276538            12.0           553.205833  
1276539            17.0           546.149412  
1276540            15.0           312.095333  
1276541            12.0           446.110000  
1276542             7.0           297.687143  

[5 rows x 37 columns]

In [16]:
cc_txn_df.columns

Index(['ssn', 'cc_num', 'acct_num', 'profile', 'trans_num', 'trans_date',
       'trans_time', 'unix_time', 'category', 'amt', 'is_fraud', 'merchant',
       'merchant_id', 'merch_long', 'merch_lat', 'first', 'last', 'gender',
       'street', 'city', 'state', 'zip', 'lat', 'long', 'city_pop', 'job',
       'dob', 'age', 'distance_from_cust', 'txn_during_night', 'trans_weekend',
       'nb_txns_1_days', 'avg_txns_amt_1_days', 'nb_txns_7_days',
       'avg_txns_amt_7_days', 'nb_txns_30_days', 'avg_txns_amt_30_days'],
      dtype='object')

### 1.2 Splitting data for atoti and AutoML

Below are the new columns for transactions and customers after computing the new features.  
We will output these features to CSV for data exploration in atoti.

In [17]:
txn_new_columns = txns_df.columns.to_list() + [
    "distance_from_cust",
    "txn_during_night",
    "trans_weekend",
    "nb_txns_1_days",
    "avg_txns_amt_1_days",
    "nb_txns_7_days",
    "avg_txns_amt_7_days",
    "nb_txns_30_days",
    "avg_txns_amt_30_days",
]

cust_new_columns = cust_df.columns.to_list() + ["age"]

We will take data from the first 3 months for exploratory data analysis and data from September for real-time simulation on atoti.

In [18]:
eda_txn_df = cc_txn_df.loc[cc_txn_df["trans_date"] < "2021-09-01"]
realtime_txn_df = cc_txn_df.loc[
    (cc_txn_df["trans_date"] >= "2021-09-01") & (cc_txn_df["trans_date"] < "2021-10-01")
]

#### 1.2.1 Export the processed data for EDA in atoti.

In [19]:
cc_txn_df[cust_new_columns].drop_duplicates().to_csv(
    f"{data_path}/post_process_customer_list.csv", index=False
)

# transaction data for ml modeling and to be loaded into atoti for eda
eda_txn_df[txn_new_columns].to_csv(
    f"{data_path}/post_process_transaction_list.csv", index=False
)

# transaction data to be exported for real-time simulation in atoti
realtime_txn_df[txn_new_columns].to_csv(f"{data_path}/realtime_txn.csv", index=False)

#### 1.2.2 Split data for machine learning and actual test on unseen data

In the example from [PyCaret tutorial](https://github.com/pycaret/pycaret/blob/master/tutorials/Binary%20Classification%20Tutorial%20Level%20Beginner%20-%20%20CLF101.ipynb), the data set is split into _data_ and _data_unseen_.  

In our case, we will use the June data for AutoML, which means that it will be further split into training and testing datasets.  
July data will be used to test the accuracy of the final model.  

In [20]:
june_data = cc_txn_df.loc[cc_txn_df["trans_date"] < "2021-07-01"]
july_aug_data = cc_txn_df.loc[
    (cc_txn_df["trans_date"] >= "2021-07-01") & (cc_txn_df["trans_date"] < "2021-09-01")
]

print(f"June: {len(june_data)} - July & August: {len(july_aug_data)}")

June: 55648 - July & August: 328520


In [21]:
july_aug_data.tail()

ssn            cc_num      acct_num  \
384163  717-56-3781   343089588484922  382840886816   
384164  282-33-3795  4486993221031082  988409340048   
384165  282-33-3795  4486993221031082  988409340048   
384166  890-25-8008   341511357666803  156690221471   
384167  488-44-6293   372091910730620  202299763857   

                               profile                         trans_num  \
384163     adults_50up_male_rural.json  726c8636d9639403c9e343b35214dc7d   
384164   adults_2550_female_urban.json  ffaee26ba5112a5b022ee5551d47063f   
384165   adults_2550_female_urban.json  b23ed48bdbf12f4556c6b76a1e00937e   
384166  young_adults_female_urban.json  1adeceacd294b7c75a33e5ba925711e6   
384167     adults_50up_male_rural.json  7c2f6c73a993dcab9d6230a8bc66c972   

       trans_date trans_time   unix_time       category     amt  ...  age  \
384163 2021-08-31   23:59:12  1630454352           home   28.55  ...   63   
384164 2021-08-31   23:59:13  1630454353  entertainment  152.87  ...   28   
384165 2021-08-31   23:59:28  1630454368  personal_care   13.05  ...   28   
384166 2021-08-31   23:59:35  1630454375  personal_care    3.59  ...   22   
384167 2021-08-31   23:59:42  1630454382           home   13.50  ...   61   

       distance_from_cust  txn_during_night  trans_weekend  nb_txns_1_days  \
384163          46.216270                 1              0             3.0   
384164          90.797907                 1              0             8.0   
384165          93.939469                 1              0             9.0   
384166          78.924718                 1              0             4.0   
384167         116.688540                 1              0             3.0   

       avg_txns_amt_1_days nb_txns_7_days avg_txns_amt_7_days nb_txns_30_days  \
384163           32.563333           31.0           57.515161           134.0   
384164           64.378750           32.0           75.802188            32.0   
384165           58.675556           33.0           73.900606            33.0   
384166           19.997500           24.0           57.830417            96.0   
384167          212.280000           19.0           82.588947            56.0   

       avg_txns_amt_30_days  
384163            61.500299  
384164            75.802188  
384165            73.900606  
384166            46.293750  
384167            61.998750  

[5 rows x 37 columns]

In [22]:
len(july_aug_data)

328520

#### 1.2.3 Select features for machine learning  

We could have used the [`ignore_features`](https://pycaret.org/classification1/) from PyCaret to ignore some data columns. However, to make it easier to see what are the features that were finally used, we perform filtered selection instead.  

In [23]:
features_list = [
    "category",
    "amt",
    "is_fraud",
    "merchant_id",
    "gender",
    "street",
    "city",
    "state",
    "zip",
    "city_pop",
    "job",
    "age",
    "distance_from_cust",
    "txn_during_night",
    "trans_weekend",
    "nb_txns_1_days",
    "avg_txns_amt_1_days",
    "nb_txns_7_days",
    "avg_txns_amt_7_days",
    "nb_txns_30_days",
    "avg_txns_amt_30_days",
]

In [24]:
data = june_data[features_list]
data_unseen = july_aug_data[features_list]

In [25]:
print(f"data: {len(data)} - data_unseen: {len(data_unseen)}")

data: 55648 - data_unseen: 328520


Reset index for the data in preparation for machine learning.

In [26]:
data.reset_index(drop=True, inplace=True)
data_unseen.reset_index(drop=True, inplace=True)

## 2 Testing autoML with PyCaret

### 2.1 Setting up environment in PyCaret

PyCaret automatically infers the data type. We can [override the inferred data type](https://pycaret.org/data-types/) adding the column names under `numeric_features` or `categorical_features`.  

Despite a smaller dataset, we still encountered memory issue. As a workaround, we add the features of high cardinality to the parameter `high_cardinality_features`.  
Note: Uncomment the `ignore_features` below if you would like to exclude the cumulative features from the machine learning model.

In [29]:
from pycaret.classification import *

clf1 = setup(
    data=data,
    target="is_fraud",
    numeric_features=["txn_during_night", "trans_weekend"],
    categorical_features=["zip", "merchant_id"],
    high_cardinality_features=["street", "city", "state", "zip", "merchant_id"],
    # ignore_features=[
    #     "nb_txns_1_days",
    #     "avg_txns_amt_1_days",
    #     "nb_txns_7_days",
    #     "avg_txns_amt_7_days",
    #     "nb_txns_30_days",
    #     "avg_txns_amt_30_days",
    # ],
)

From the table, we can see that the number of features increased from 21 to 492 due to categorical encoding:
- The Original Data	(55648, 21)
- Transformed Train Set	(38953, 492)
- Transformed Test Set	(16695, 492)

### 2.2 Comparing all models

The goal is to be able to detect potential fraud and enable the bank to investigate or alert the consumer of suspicious transactions. Therefore, we will choose a few models with the highest F1 score, AUC and Recall for further comparison. 

In [30]:
best_model = compare_models()

In [31]:
best_model

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
               random_state=6478, reg_alpha=0.0, reg_lambda=0.0, silent='warn',
               subsample=1.0, subsample_for_bin=200000, subsample_freq=0)

### 2.3 Creating models

Let's create the model for a few of those with highest F1 score and AUC. In fraud detection, we aim to raise alert to the credit cardholder of suspicious transactions. Hence, we will take the model that scored well for both AUC and recall.  

#### 2.3.1 Decision Tree Classifier

In [32]:
dt = create_model("dt")

#### 2.3.2 Extra Trees Classifier

In [33]:
et = create_model("et")

#### 2.3.3 Light Gradient Boosting Machine

In [34]:
lgbm = create_model("lightgbm")

### 2.4 Tune models

`create_model()` trains the model using default hyperparameters. We will use `tune_model()` function to tune the hyperparameters.  

In [35]:
tuned_dt = tune_model(dt)

In [36]:
tuned_et = tune_model(et)

In [37]:
tuned_lgbm = tune_model(lgbm)

### 2.5 Evaluating models

The `evaluate_model()` function displays a user interface for all the available plots for a given model. Let's see the performance for each of the models.

In [38]:
evaluate_model(tuned_dt)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

In [39]:
evaluate_model(tuned_et)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

In [40]:
evaluate_model(tuned_lgbm)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

### 2.6 Predict on test/hold-out sample  

We perform one final check by predicting the test/hold-out set before we finalize the model. We saw earlier in section 2.1, we have 16,695 samples for the test data:  
__Transformed Test Set: (16695, 491)__

In [41]:
predict_model(tuned_dt);

In [42]:
predict_model(tuned_et);

In [43]:
predict_model(tuned_lgbm);

From the evaluation, we see that for the Extra Trees Classifier, the recall, precision and other metrics becomes 0. It also scores poorly for the AUC.  
We can probably drop the model from our evaluation.

### 2.7 Finalize model for deployment  

Lastly, we proceed to train the model on the complete dataset with the `finalize_model()`. This should lead to the best model for use in making prediction on new and unseen data.

In [44]:
final_dt = finalize_model(tuned_dt)

In [45]:
final_et = finalize_model(tuned_et)

In [46]:
final_lgbm = finalize_model(tuned_lgbm)

In [47]:
print(final_dt)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=8, max_features=1.0, max_leaf_nodes=None,
                       min_impurity_decrease=0, min_impurity_split=None,
                       min_samples_leaf=6, min_samples_split=10,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=6478, splitter='best')


In [48]:
print(final_lgbm)

LGBMClassifier(bagging_fraction=0.4, bagging_freq=2, boosting_type='gbdt',
               class_weight=None, colsample_bytree=1.0, feature_fraction=0.6,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=81, min_child_weight=0.001, min_split_gain=0.7,
               n_estimators=240, n_jobs=-1, num_leaves=70, objective=None,
               random_state=6478, reg_alpha=0.4, reg_lambda=0.005,
               silent='warn', subsample=1.0, subsample_for_bin=200000,
               subsample_freq=0)


In [49]:
print(final_et)

ExtraTreesClassifier(bootstrap=True, ccp_alpha=0.0, class_weight={},
                     criterion='entropy', max_depth=3, max_features=1.0,
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.4, min_impurity_split=None,
                     min_samples_leaf=6, min_samples_split=5,
                     min_weight_fraction_leaf=0.0, n_estimators=110, n_jobs=-1,
                     oob_score=False, random_state=6478, verbose=0,
                     warm_start=False)


### 2.8 Predict on unseen data  

We have previously used June data for ML modelling and put aside the July and August data in section 1.2.2.  
Now, we will use the unseen July and Aug data to test out the finalized model.  

#### 2.8.1 Decision Tree Classifier

In [50]:
prediction_output_col = ["trans_num", "Label", "Score"]

`Label` and `Score` columns are added onto the data_unseen set:  
- Label is the prediction  
- score is the probability of the prediction.  

In [51]:
unseen_predictions = predict_model(final_dt, data=data_unseen)
unseen_predictions.head()

category     amt  is_fraud  merchant_id gender  \
0   shopping_pos   14.11         0        81458      F   
1    grocery_pos   75.52         0       490762      M   
2    grocery_pos  136.23         0       635517      F   
3  gas_transport   66.94         0        19638      F   
4  entertainment  116.85         0        23622      M   

                         street              city state    zip  city_pop  ...  \
0   7820 Wilson Tunnel Apt. 795           Andrews    SC  29510     11060  ...   
1          2137 Serrano Station  North Charleston    SC  29420     71702  ...   
2           4656 Karen Parkways              Elmo    MT  59915       403  ...   
3          74769 Anderson Plain           Oaktown    IN  47561      1588  ...   
4  318 Brenda Terrace Suite 001           Beverly    WV  26253      3490  ...   

  txn_during_night  trans_weekend  nb_txns_1_days  avg_txns_amt_1_days  \
0                1              0             1.0                14.11   
1                1              0             1.0                75.52   
2                1              0             1.0               136.23   
3                1              0             1.0                66.94   
4                1              0             1.0               116.85   

   nb_txns_7_days  avg_txns_amt_7_days  nb_txns_30_days  avg_txns_amt_30_days  \
0            22.0            62.525455             33.0             55.510909   
1            27.0            50.487407             69.0             73.280870   
2            34.0            88.363529            127.0             69.842756   
3            10.0            28.290000             60.0             30.620000   
4            19.0            66.212105             92.0             65.748152   

   Label   Score  
0      0  0.9993  
1      0  0.9993  
2      0  0.9993  
3      0  1.0000  
4      0  0.9993  

[5 rows x 23 columns]

It's nice that the predicted results are concatenated to the original dataset while all the transformations are automatically performed in the background.  

However, in section 1.2.3, we have selected only the features that we needed for the ML because we have a lot of features that we decided not necessary.  
Hence, we will do another merge back to the original data frame as we needed the `trans_num` for our use in atoti later on.  

In [52]:
july_aug_prediction_dt = july_aug_data.merge(unseen_predictions)
july_aug_prediction_dt.head(3)

ssn            cc_num      acct_num                        profile  \
0  547-69-7793  6534665167391525  693931531856  adults_50up_female_urban.json   
1  191-61-6759  4166219838226478  181679512065    adults_2550_male_urban.json   
2  027-46-7326  3513900304593051  920573624386  adults_50up_female_rural.json   

                          trans_num trans_date trans_time   unix_time  \
0  e14e1a05c8da5e3f3da457e6eb84d825 2021-07-01   00:00:18  1625097618   
1  a3c0100cb4cfb7a5c4ffc90faaf7a6d8 2021-07-01   00:01:15  1625097675   
2  9952d8bafe837216db5134d988d5144a 2021-07-01   00:02:12  1625097732   

       category     amt  ...  txn_during_night trans_weekend  nb_txns_1_days  \
0  shopping_pos   14.11  ...                 1             0             1.0   
1   grocery_pos   75.52  ...                 1             0             1.0   
2   grocery_pos  136.23  ...                 1             0             1.0   

   avg_txns_amt_1_days  nb_txns_7_days avg_txns_amt_7_days nb_txns_30_days  \
0                14.11            22.0           62.525455            33.0   
1                75.52            27.0           50.487407            69.0   
2               136.23            34.0           88.363529           127.0   

  avg_txns_amt_30_days Label   Score  
0            55.510909     0  0.9993  
1            73.280870     0  0.9993  
2            69.842756     0  0.9993  

[3 rows x 39 columns]

This step is purely to output the prediction for loading into atoti. We will use atoti to gather some insights on the impact of the prediction on the various business metrics.

In [53]:
july_aug_prediction_dt[prediction_output_col].to_csv(
    f"{data_path}/july_aug_prediction_full_dt.csv", index=False
)

#### 2.8.2 Extra Trees Classifier

For completeness, we will see the performance of this model despite its poor performance during the training.

In [54]:
unseen_predictions_et = predict_model(final_et, data=data_unseen)
unseen_predictions_et.head()

category     amt  is_fraud  merchant_id gender  \
0   shopping_pos   14.11         0        81458      F   
1    grocery_pos   75.52         0       490762      M   
2    grocery_pos  136.23         0       635517      F   
3  gas_transport   66.94         0        19638      F   
4  entertainment  116.85         0        23622      M   

                         street              city state    zip  city_pop  ...  \
0   7820 Wilson Tunnel Apt. 795           Andrews    SC  29510     11060  ...   
1          2137 Serrano Station  North Charleston    SC  29420     71702  ...   
2           4656 Karen Parkways              Elmo    MT  59915       403  ...   
3          74769 Anderson Plain           Oaktown    IN  47561      1588  ...   
4  318 Brenda Terrace Suite 001           Beverly    WV  26253      3490  ...   

  txn_during_night  trans_weekend  nb_txns_1_days  avg_txns_amt_1_days  \
0                1              0             1.0                14.11   
1                1              0             1.0                75.52   
2                1              0             1.0               136.23   
3                1              0             1.0                66.94   
4                1              0             1.0               116.85   

   nb_txns_7_days  avg_txns_amt_7_days  nb_txns_30_days  avg_txns_amt_30_days  \
0            22.0            62.525455             33.0             55.510909   
1            27.0            50.487407             69.0             73.280870   
2            34.0            88.363529            127.0             69.842756   
3            10.0            28.290000             60.0             30.620000   
4            19.0            66.212105             92.0             65.748152   

   Label   Score  
0      0  0.9795  
1      0  0.9795  
2      0  0.9795  
3      0  0.9795  
4      0  0.9795  

[5 rows x 23 columns]

In [55]:
july_aug_prediction_et = july_aug_data.merge(unseen_predictions_et)
july_aug_prediction_et.head(3)

ssn            cc_num      acct_num                        profile  \
0  547-69-7793  6534665167391525  693931531856  adults_50up_female_urban.json   
1  191-61-6759  4166219838226478  181679512065    adults_2550_male_urban.json   
2  027-46-7326  3513900304593051  920573624386  adults_50up_female_rural.json   

                          trans_num trans_date trans_time   unix_time  \
0  e14e1a05c8da5e3f3da457e6eb84d825 2021-07-01   00:00:18  1625097618   
1  a3c0100cb4cfb7a5c4ffc90faaf7a6d8 2021-07-01   00:01:15  1625097675   
2  9952d8bafe837216db5134d988d5144a 2021-07-01   00:02:12  1625097732   

       category     amt  ...  txn_during_night trans_weekend  nb_txns_1_days  \
0  shopping_pos   14.11  ...                 1             0             1.0   
1   grocery_pos   75.52  ...                 1             0             1.0   
2   grocery_pos  136.23  ...                 1             0             1.0   

   avg_txns_amt_1_days  nb_txns_7_days avg_txns_amt_7_days nb_txns_30_days  \
0                14.11            22.0           62.525455            33.0   
1                75.52            27.0           50.487407            69.0   
2               136.23            34.0           88.363529           127.0   

  avg_txns_amt_30_days Label   Score  
0            55.510909     0  0.9795  
1            73.280870     0  0.9795  
2            69.842756     0  0.9795  

[3 rows x 39 columns]

In [56]:
july_aug_prediction_et[prediction_output_col].to_csv(
    f"{data_path}/july_aug_prediction_full_et.csv", index=False
)

#### 2.8.3 Light Gradient Boosting Machine

During the model compare, LGBM has the highest AUC. Let's see if it do as well for the unseen data prediction.

In [57]:
unseen_predictions_lgbm = predict_model(final_lgbm, data=data_unseen)
unseen_predictions_lgbm.head()

category     amt  is_fraud  merchant_id gender  \
0   shopping_pos   14.11         0        81458      F   
1    grocery_pos   75.52         0       490762      M   
2    grocery_pos  136.23         0       635517      F   
3  gas_transport   66.94         0        19638      F   
4  entertainment  116.85         0        23622      M   

                         street              city state    zip  city_pop  ...  \
0   7820 Wilson Tunnel Apt. 795           Andrews    SC  29510     11060  ...   
1          2137 Serrano Station  North Charleston    SC  29420     71702  ...   
2           4656 Karen Parkways              Elmo    MT  59915       403  ...   
3          74769 Anderson Plain           Oaktown    IN  47561      1588  ...   
4  318 Brenda Terrace Suite 001           Beverly    WV  26253      3490  ...   

  txn_during_night  trans_weekend  nb_txns_1_days  avg_txns_amt_1_days  \
0                1              0             1.0                14.11   
1                1              0             1.0                75.52   
2                1              0             1.0               136.23   
3                1              0             1.0                66.94   
4                1              0             1.0               116.85   

   nb_txns_7_days  avg_txns_amt_7_days  nb_txns_30_days  avg_txns_amt_30_days  \
0            22.0            62.525455             33.0             55.510909   
1            27.0            50.487407             69.0             73.280870   
2            34.0            88.363529            127.0             69.842756   
3            10.0            28.290000             60.0             30.620000   
4            19.0            66.212105             92.0             65.748152   

   Label   Score  
0      0  0.9990  
1      0  0.9998  
2      0  0.9998  
3      0  0.9999  
4      0  0.9998  

[5 rows x 23 columns]

In [58]:
july_aug_prediction_lgbm = july_aug_data.merge(unseen_predictions_lgbm)
july_aug_prediction_lgbm.head(3)

ssn            cc_num      acct_num                        profile  \
0  547-69-7793  6534665167391525  693931531856  adults_50up_female_urban.json   
1  191-61-6759  4166219838226478  181679512065    adults_2550_male_urban.json   
2  027-46-7326  3513900304593051  920573624386  adults_50up_female_rural.json   

                          trans_num trans_date trans_time   unix_time  \
0  e14e1a05c8da5e3f3da457e6eb84d825 2021-07-01   00:00:18  1625097618   
1  a3c0100cb4cfb7a5c4ffc90faaf7a6d8 2021-07-01   00:01:15  1625097675   
2  9952d8bafe837216db5134d988d5144a 2021-07-01   00:02:12  1625097732   

       category     amt  ...  txn_during_night trans_weekend  nb_txns_1_days  \
0  shopping_pos   14.11  ...                 1             0             1.0   
1   grocery_pos   75.52  ...                 1             0             1.0   
2   grocery_pos  136.23  ...                 1             0             1.0   

   avg_txns_amt_1_days  nb_txns_7_days avg_txns_amt_7_days nb_txns_30_days  \
0                14.11            22.0           62.525455            33.0   
1                75.52            27.0           50.487407            69.0   
2               136.23            34.0           88.363529           127.0   

  avg_txns_amt_30_days Label   Score  
0            55.510909     0  0.9990  
1            73.280870     0  0.9998  
2            69.842756     0  0.9998  

[3 rows x 39 columns]

In [59]:
july_aug_prediction_lgbm[prediction_output_col].to_csv(
    f"{data_path}/july_aug_prediction_full_lgbm.csv", index=False
)

We will use the `check_metric` function from the `pycaret.utils` to get the AUC of the run. We can change and compare the metric to recall if preferred.  
It seems that LGBM still has the best AUC. But generally, the AUC scores are lower than what we have on the predict_mode stage in section 2.5.  

We shall see if the same holds true when we perform real-time streaming and ML evaluation in atoti.

In [60]:
from pycaret.utils import check_metric

check_metric(unseen_predictions["is_fraud"], unseen_predictions["Label"], metric="AUC")

0.9268

In [61]:
check_metric(
    unseen_predictions_et["is_fraud"], unseen_predictions_et["Label"], metric="AUC"
)

0.5

In [62]:
check_metric(
    unseen_predictions_lgbm["is_fraud"], unseen_predictions_lgbm["Label"], metric="AUC"
)

0.9285

### 2.9 Save trained model for real-time fraud detection with atoti

It takes a long time to train the models. Hence, it's great that we can save and use the model without having to repeat these all the time.  

In [63]:
save_model(final_dt, "Final DT Model 20211130")

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=None,
          steps=[('dtypes',
                  DataTypes_Auto_infer(categorical_features=['zip',
                                                             'merchant_id'],
                                       display_types=True, features_todrop=[],
                                       id_columns=[],
                                       ml_usecase='classification',
                                       numerical_features=['txn_during_night',
                                                           'trans_weekend'],
                                       target='is_fraud', time_features=[])),
                 ('imputer',
                  Simple_Imputer(categorical_strategy='not_available',
                                 fill_value_c...
                 ('dfs', 'passthrough'), ('pca', 'passthrough'),
                 ['trained_model',
                  DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None,
                                         crit

In [64]:
save_model(final_et, "Final ET Model 20211130")

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=None,
          steps=[('dtypes',
                  DataTypes_Auto_infer(categorical_features=['zip',
                                                             'merchant_id'],
                                       display_types=True, features_todrop=[],
                                       id_columns=[],
                                       ml_usecase='classification',
                                       numerical_features=['txn_during_night',
                                                           'trans_weekend'],
                                       target='is_fraud', time_features=[])),
                 ('imputer',
                  Simple_Imputer(categorical_strategy='not_available',
                                 fill_value_c...
                  ExtraTreesClassifier(bootstrap=True, ccp_alpha=0.0,
                                       class_weight={}, criterion='entropy',
                                       max_depth=3, max_features=1.0,
   

In [65]:
save_model(final_lgbm, "Final LGBM Model 20211130")

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=None,
          steps=[('dtypes',
                  DataTypes_Auto_infer(categorical_features=['zip',
                                                             'merchant_id'],
                                       display_types=True, features_todrop=[],
                                       id_columns=[],
                                       ml_usecase='classification',
                                       numerical_features=['txn_during_night',
                                                           'trans_weekend'],
                                       target='is_fraud', time_features=[])),
                 ('imputer',
                  Simple_Imputer(categorical_strategy='not_available',
                                 fill_value_c...
                                 colsample_bytree=1.0, feature_fraction=0.6,
                                 importance_type='split', learning_rate=0.1,
                                 max_depth=-1, min_child_samples=8